### ISSUE TITLE
Allow drive-by contributors to login using Github-without-2FA

### ISSUE THREAD

**Issue by hmitsch** [mozilla-iam/auth0-deploy#166]
In https://github.com/mozilla-iam/auth0-deploy/pull/145 we forced 2FA on people using Github as their authentication provider.

The recent Discourse thread https://discourse.mozilla.org/t/github-login-issue/23730 shows that this locks out people from participating in Discourse conversations who are not holding any Mozilla IAM _Access Groups_ (because they do not even have a Mozillians.org profile). For sake of simplicity, let's call this group _drive-by contributors_.

As Github signals whether or not 2FA is being used, we could change towards the following business rules:

**Hierachy for _Most Secure_ Login:**
> Mozilla LDAP+MFA > Github+2FA > Github|Google > Passwordless Email

_Access Groups_ would only be relayed for MFA/2FA login methods (i.e. LDAP and Github+2FA).

**Conclusion:**
Such a change would allow _drive-by contributors_ to use Github as authentication provider and have a username+password login method on Mozilla sites such as Discourse.

**Comment by LeoMcA** [comment]
> Hierachy for Most Secure Login:
>>    Mozilla LDAP+MFA > Github+2FA > Github|Google > Passwordless Email

I don't want to muddy the waters in this issue, but after making this change is Passwordless less secure than non-2FA Github/Google? We definitely want to encourage users to use non-2FA Github/Google over Passwordless, but should we be forcing them? I know there's a very vocal group of users on Discourse who really don't want to be forced to use a third-party to authenticate.

I wonder if the hierarchy should become:

    Mozilla LDAP + MFA > Github + 2FA > Github | Google | Passwordless

**Comment by gdestuynder** [comment]
Accounts with a single authentication method, which is generally a password are less safe than accounts with at least 2 authentication methods, as in the identity is easier to be compromised (1 factor vs 2 factors). See also https://infosec.mozilla.org/fundamentals/rationales.html#mfa

I would generally recommend not having accounts where we are accountable for identity's integrity without 2FA or a strong 1FA (passwordless and password-based are not strong 1FA).

Out of curiosity, can you quantity the % of demand to not use 2FA by Discourse users? (the percentage is much more interesting than "how vocal")

**Comment by LeoMcA** [comment]
> Out of curiosity, can you quantity the % of demand to not use 2FA by Discourse users? (the percentage is much more interesting than "how vocal")

Of the **1978** users who have logged in since we started saving user_ids in Discourse (on Nov 1st):

| Method | Users | % |
| - | - | - |
| LDAP | 95 | 5 |
| GitHub | 447 | 23 |
| Google | 841 | 43 |
| Passwordless | 595 | 30 |

That means, of those users, **73%** are using a 1FA method. I'd say that's pretty high demand to not use 2FA ;)

I should add a disclaimer, however, that these statistics may be biased against very active users, as we only forced a logout on Dec 20th, so a number of users using 2FA methods may have remained logged in until that point (and have remain logged out past that point, due to the holidays), so we haven't saved their user_id yet. I can query the db again in a couple of weeks to get a more up-to-date picture, if need-be.

**Comment by jeffbryner** [comment]
I agree with @gdestuynder that in 2018 the bar for contributions to projects like Mozilla that reach hundreds of millions of users is 2FA at a minimum. There are plenty of tools (pastebin, email lists, etc) that don't require auth and don't pretend to vouch for integrity that can be used for 'drive-by' discussions/contributors who are unable/unwilling to reach the minimum requirements.

**Comment by hmitsch** [comment]
Excellent point!
I think we all agree that code contributions to Firefox or any related ecosystem product need to be on 2FA at a minimum.

There seems to be a grey zone which requires a case-by-case decision of _valuing user integrity over authentication convenience_. Let me provide two examples:
* **Discourse**
Discourse is used as discussion forum and mailing list by various Mozilla communities (e.g. Add-ons, Testpilot, Emerging Technologies, and many others). In 2018 we intend to evaluate if SUMO and various regional Mozilla communities also want to migrate to Discourse.
* **MDN**
MDN currently uses it's own Github authentication to login to [developer.mozilla.org](https://developer.mozilla.org). It does not force users to have 2FA enabled. Data shows that many MDN contributions are  by one-time contributors. This means there is a programmatic requirement to have the most seamless and convenient login experience.

Would we consider these two use cases to fall into the category _where we are accountable for identity's integrity_? If yes, then we should stick with the 2FA requirement until we find new data that would reorient our perspective. If no, we should make the investment to allow Github-without-2FA login for these services.

**Comment by edmorley** [comment]
Hopefully #215 will reduce user confusion (eg as seen in mozilla/discourse#123) slightly in the meantime.

**Comment by edmorley** [comment]
(mozilla-iam/sso-dashboard#215 that is)
